### PR TITLE
Decryption

### LINKED ISSUE #49: Decryption support
I have been working on adding support for decrypting messages and if it is ok I would like to submit a PR with this functionality once I am done.

I have an initial version that _almost_ works at https://github.com/furuholm/libsignal-protocol-rs/tree/decryption. However, it turns out that the decryption code in the underlying C library requires the mutex implementation to be reentrant which `parking_lot::RawMutex` is not. When replacing `RawMutex` with `ReentrantMutex` I get the following error message

```
error[E0277]: the type `std::cell::UnsafeCell<usize>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   --> libsignal-protocol/src/context.rs:410:17
    |
410 |         let _ = std::panic::catch_unwind(|| {
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^ `std::cell::UnsafeCell<usize>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
    |
    = help: within `&context::State`, the trait `std::panic::RefUnwindSafe` is not implemented for `std::cell::UnsafeCell<usize>`
    = note: required because it appears within the type `std::cell::Cell<usize>`
    = note: required because it appears within the type `lock_api::remutex::RawReentrantMutex<parking_lot::raw_mutex::RawMutex, parking_lot::remutex::RawThreadId>`
    = note: required because it appears within the type `lock_api::remutex::ReentrantMutex<parking_lot::raw_mutex::RawMutex, parking_lot::remutex::RawThreadId, ()>`
    = note: required because it appears within the type `context::State`
    = note: required because it appears within the type `&context::State`
    = note: required because of the requirements on the impl of `std::panic::UnwindSafe` for `&&context::State`
    = note: required because it appears within the type `[closure@libsignal-protocol/src/context.rs:410:42: 413:10 state:&&context::State, level:&log::Level, message:&&str]`
 ```

 Just like the message says it is caused by `RawReentrantMutex` containing a `cell::Cell` which is not `RefUnwindSafe`.

 Any ideas on how to solve this?

### PR DESCRIPTION
**Pull request by furuholm**
Solves #49

### CODE REVIEW

File: libsignal-protocol/tests/libsignal-protocol-c-tests.rs [hunk 1]
```diff
@@ -400,3 +411,124 @@ fn test_optional_one_time_pre_key() {
     let has_pre_key_id = incoming_message.has_pre_key_id();
     assert!(!has_pre_key_id);
 }
+
+#[test]
+fn test_decrypt() {
```
Comment (Michael-F-Bryan) [inline libsignal-protocol/tests/libsignal-protocol-c-tests.rs]: Is this a port of one of the original `libsignal-protocol-c` tests?

File: libsignal-protocol/tests/libsignal-protocol-c-tests.rs [hunk 2]
```diff
@@ -400,3 +411,124 @@ fn test_optional_one_time_pre_key() {
...
+    let bob_store = sig::store_context(
+        &ctx,
+        InMemoryPreKeyStore::default(),
+        InMemorySignedPreKeyStore::default(),
+        InMemorySessionStore::default(),
+        InMemoryIdentityKeyStore::new(registration_id, &bob_identity_key_pair),
+    )
+    .unwrap();
+
+    let bob_local_registration_id = bob_store.registration_id().unwrap();
```
Comment (Michael-F-Bryan) [inline libsignal-protocol/tests/libsignal-protocol-c-tests.rs]: What's the difference between `bob_local_registration_id` and `registration_id`? Are they actually the same ID?

File: libsignal-protocol/tests/libsignal-protocol-c-tests.rs [hunk 3]
```diff
@@ -400,3 +411,124 @@ fn test_optional_one_time_pre_key() {
...
+    let alice_session_cipher =
+        sig::SessionCipher::new(&ctx, &alice_store, &bob_address).unwrap();
+
+    // Let Alice send a message to bob
+    let msg = "Hello bob!";
+    let outgoing_message =
+        alice_session_cipher.encrypt(msg.as_bytes()).unwrap();
+
+    // Convert to an incoming message (this is technically a downcast from
+    // CiphertextMessage to PreKeySignalMessage)
```
Comment (Michael-F-Bryan) [inline libsignal-protocol/tests/libsignal-protocol-c-tests.rs]: Do we have any guarantee that `outgoing_message` will always be `PreKeySignalMessage`? If so, it might be easier if the `SessionCipher::encrypt()` method returns a `PreKeySignalMessage` directly seeing as that's the more specific "class".

File: libsignal-protocol/tests/libsignal-protocol-c-tests.rs [hunk 4]
```diff
@@ -400,3 +411,124 @@ fn test_optional_one_time_pre_key() {
...
+        sig::SessionCipher::new(&ctx, &bob_store, &alice_address).unwrap();
+
+    // Decrypt message to bob
+    let decrypted_msg = bob_session_cipher
+        .decrypt_pre_key_message(&incoming_message)
+        .unwrap();
+    assert_eq!(msg, std::str::from_utf8(decrypted_msg.as_slice()).unwrap());
+
+    // Let bob send a message to Alice
+    let msg = "Hi Alice!";
```
Comment (Michael-F-Bryan) [inline libsignal-protocol/tests/libsignal-protocol-c-tests.rs]: Lol, we needed about 100 lines of setup and key exchanges when the actual interesting bit (sending a message) takes less than 10 lines.

Is sending messages between clients always going to be this annoying/tedious/stateful, or will higher-level applications be able simplify the process later on?

File: libsignal-protocol/src/session_cipher.rs [hunk 5]
```diff
@@ -66,6 +66,44 @@ impl SessionCipher {
...
+    pub fn decrypt_pre_key_message(
+        &self,
+        message: &PreKeySignalMessage,
+    ) -> Result<Buffer, Error> {
+        unsafe {
+            let mut buffer = ptr::null_mut();
+            sys::session_cipher_decrypt_pre_key_signal_message(
+                self.raw,
+                message.raw.as_ptr(),
+                ptr::null_mut(),
```
[the remark this answers is not in the document]
Comment (furuholm) [reply libsignal-protocol/src/session_cipher.rs]: Is this something we should do now? What use case do you see for this?

File: libsignal-protocol/src/messages/signal_message.rs [hunk 6]
```diff
@@ -119,4 +119,33 @@ impl From<SignalMessage> for CiphertextMessage {
...
+            let raw = Raw::copied_from(raw as *mut sys::signal_message);
+            Ok(SignalMessage {
+                raw,
+                _ctx: Rc::clone(&ctx.0),
+            })
+        }
+    }
+
+    fn serialize(&self) -> Result<Buffer, failure::Error> {
+        unimplemented!()
```
[the remark this answers is not in the document]
Comment (furuholm) [reply libsignal-protocol/src/messages/signal_message.rs]: I don't think this will ever be implemented as it is not supported by the underlying library. 

I would say that serialization is typically done as follows:

1. Encrypt message with `SessionCipher::encrypt(&self, message: &[u8]) -> Result<CiphertextMessage, Error>`
2. Serialize the returned `CiphertextMessage`
3. Send serialized message together with stuff like destination and message type (pre-key message, message, sender key...) to server.

Deserialization:

1. Fetch message envelope from server.
2. Check said envelope for type of message.
3. Depending on type, invoke `PreKeySignalMessage::deserialize` or `SignalMessage::deserialize`.
4. Depending on type, invoke either `SessionCipher::decrypt_pre_key_message` or `SessionCipher::decrypt_message`

This is why `CiphertextMessage` only implements `serialize` and `PreKeySignalMessage` and `SignalMessage` both implement `deserialize`.

It could be argued that the Rust API should operate at a higher level, but I probably think it is better to stick to the same semantics as the C library.

File: libsignal-protocol/src/messages/signal_message.rs [hunk 7]
```diff
@@ -119,4 +119,33 @@ impl From<SignalMessage> for CiphertextMessage {
...
+            let res = sys::signal_message_deserialize(
+                &mut raw,
+                data.as_ptr(),
+                data.len(),
+                ctx.raw(),
+            );
+            if res != 0 {
+                return Err(failure::err_msg("Unable to deserialize buffer"));
+            }
+            // safety: the successful invocation of signal_message_deserialize
```
[the remark this answers is not in the document]
Comment (furuholm) [reply libsignal-protocol/src/messages/signal_message.rs]: You are right. Fixed in eddfe7b02. I should write a test for this though.

File: libsignal-protocol/src/messages/signal_message.rs [hunk 8]
```diff
@@ -119,4 +119,33 @@ impl From<SignalMessage> for CiphertextMessage {
...
+        unsafe {
+            let mut raw = ptr::null_mut();
+
+            let res = sys::signal_message_deserialize(
+                &mut raw,
+                data.as_ptr(),
+                data.len(),
+                ctx.raw(),
+            );
+            if res != 0 {
```
[the remark this answers is not in the document]
Comment (furuholm) [reply libsignal-protocol/src/messages/signal_message.rs]: Fixed in ffd7c4db2.

File: libsignal-protocol/src/messages/pre_key_signal_message.rs [hunk 9]
```diff
@@ -135,4 +135,33 @@ impl From<PreKeySignalMessage> for CiphertextMessage {
...
+impl Serializable for PreKeySignalMessage {
+    fn deserialize(ctx: &Context, data: &[u8]) -> Result<Self, failure::Error> {
+        unsafe {
+            let mut raw = ptr::null_mut();
+
+            let res = sys::pre_key_signal_message_deserialize(
+                &mut raw,
+                data.as_ptr(),
+                data.len(),
+                ctx.raw(),
```
[the remark this answers is not in the document]
Comment (furuholm) [reply libsignal-protocol/src/messages/pre_key_signal_message.rs]: `signal_message` contains a pointer to the global context, so `deserialize` needs the context to create a new message.
`serialize` does not require a `Context`.

File: libsignal-protocol/tests/libsignal-protocol-c-tests.rs (hunk 1 above)
Comment (furuholm) [reply libsignal-protocol/tests/libsignal-protocol-c-tests.rs]: No it is not. I realize now that the other tests have corresponding tests in the `libsignal-protocol-c`. I think I could rewrite this to match `test_basic_pre_key_v3`. That test seems to match `test_decrypt` quite closely.

File: libsignal-protocol/tests/libsignal-protocol-c-tests.rs (hunk 2 above)
Comment (furuholm) [reply libsignal-protocol/tests/libsignal-protocol-c-tests.rs]: They are. Fixed in 34149a116.

File: libsignal-protocol/tests/libsignal-protocol-c-tests.rs (hunk 3 above)
Comment (furuholm) [reply libsignal-protocol/tests/libsignal-protocol-c-tests.rs]: No, it won't always be `PreKeySignalMessage`. A few lines below in the test a `SignalMessage` is returned.

To be more precise: it will be a `PreKeySignalMessage` until a message reply has been received. This means that if Alice initiates the conversation with Bob she will never receive a `PreKeySignalMessage` as the first message she gets from Bob will be a reply. 
Alice will send `PreKeySignalMessage`s to Bob until she receives a message from Bob. After this, any subsequent messages sent to Bob will be `SignalMessage`s.

File: libsignal-protocol/tests/libsignal-protocol-c-tests.rs (hunk 4 above)
Comment (furuholm) [reply libsignal-protocol/tests/libsignal-protocol-c-tests.rs]: Once the communication between two clients have been established there is not that much code that is involved in sending and receiving messages.
Most of this code will be invoked during device setup and during messaging session setup. For example:
- At device setup identity key and pre keys will be generated and the server will generate registration id that the client stores.
- The pre key will be fetched from the server and processed once the user decides to start a communication with another user.

File: libsignal-protocol/src/messages/signal_message.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (furuholm) [reply libsignal-protocol/src/messages/signal_message.rs]: This would be tested if we port ` test_basic_pre_key_v3` (see https://github.com/Michael-F-Bryan/libsignal-protocol-rs/pull/54#discussion_r363025361)

File: libsignal-protocol/tests/libsignal-protocol-c-tests.rs (hunk 1 above)
Comment (Michael-F-Bryan) [reply libsignal-protocol/tests/libsignal-protocol-c-tests.rs]: Na, that's alright. I think most of the other tests were copied from C because that was the most readily accessible source of code which uses this library and shows what it expects to happen.

TL;DR: There's no reason for the correspondence other than Michael being lazy and copying existing tests :P

File: libsignal-protocol/tests/libsignal-protocol-c-tests.rs (hunk 4 above)
Comment (Michael-F-Bryan) [reply libsignal-protocol/tests/libsignal-protocol-c-tests.rs]: That makes a lot of sense. So it seems tedious because we're doing all the setup in one place when it would normally be done once per conversation (i.e. Alice starts talking to Bob) then a bit more before sending a message.

File: libsignal-protocol/src/messages/pre_key_signal_message.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Michael-F-Bryan) [reply libsignal-protocol/src/messages/pre_key_signal_message.rs]: Damn, that's annoying...

File: libsignal-protocol/src/session_cipher.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Michael-F-Bryan) [reply libsignal-protocol/src/session_cipher.rs]: I was thinking it might be usable for things like logging, middleware, or progress monitoring.

I doubt we'd need it now, but it's nice to know this sort of thing exists for later.

File: libsignal-protocol/src/messages/signal_message.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Michael-F-Bryan) [reply libsignal-protocol/src/messages/signal_message.rs]: What about situations where you want to use example messages as test fixtures? Under those circumstances it would be beneficial to implement both `serialize()` (for saving to disk in the first place) and `deserialize()` (for reading from disk during test setup)... But I'm guessing if it's not implemented by the C library our hands are tied, aren't they?

**Comment by Michael-F-Bryan** [comment]
Thanks for all the hard work @furoholm! 

In terms of functionality I'm pretty happy with this PR. I've got a couple more questions, but they're more along the lines of high-level design questions (e.g. implementing both `serialize()` and `deserialize()`) and not directly related to the PR.

Is there anything else you'd like to add, or should I just the merge button?

File: libsignal-protocol/src/messages/signal_message.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (furuholm) [reply libsignal-protocol/src/messages/signal_message.rs]: Could you provide a more concrete example of what this would look like and why it would be something we would want?

File: libsignal-protocol/tests/libsignal-protocol-c-tests.rs (hunk 4 above)
Comment (furuholm) [reply libsignal-protocol/tests/libsignal-protocol-c-tests.rs]: Exactly

File: libsignal-protocol/src/messages/signal_message.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (furuholm) [reply libsignal-protocol/src/messages/signal_message.rs]: Some ramblings:

Regarding serialize: 
`libsignal-protocol-rs` doesn't provide an API to create `PreKeySignalMessage` or `SignalMessage` explicitly, so what would trigger the need to convert a `CiphertextMessage` to a `PreKeySignalMessage` or a `SignalMessage` before serializing it?

Regarding deserialize: 
If it wasn't for the fact that the server provides the type in the message envelope I think deserialization of `CiphertextMessage` would make more sense. To do it I think we would have to try to deserialize to `PreKeySignalMessage` or `SignalMessage` and see what succeeds and then upcast to `CiphertextMessage`. This would then have to be converted to a `PreKeySignalMessage` or `SignalMessage` before being useful. 

_If we were to implement it (which I am not suggesting) it could maybe make sense to return an `enum` which would either contain `PreKeySignalMessage` or `SignalMessage`. Maybe `CiphertextMessage` would be that enum?_

**Comment by furuholm** [comment]
I'm not in a hurry to get this merged so I think we could settle the design discussion regarding `serialize` and `deserialize` before merging. Would also like to add serialization/deserialization to the test suite and rebase before merging.

File: libsignal-protocol/tests/libsignal-protocol-c-tests.rs (hunk 1 above)
Comment (furuholm) [reply libsignal-protocol/tests/libsignal-protocol-c-tests.rs]: I did rewrite the test and rename it `test_basic_pre_key_v3` in d5bd540b1. It is not a one-to-one mapping as the `libsignal-protocol-c` operates at a lower level, but the basic structure should (roughly) match.

This test also includes serialization and deserialization.

**Comment by furuholm** [comment]
Rebase done. Also dd0641b62 includes testing of serialization and deserialization.

**Comment by furuholm** [comment]
The test failed in Travis. Will check tomorrow...

**Comment by furuholm** [comment]
Seems like decryption fails with `--features="crypto-openssl"` at this line:

https://github.com/sfackler/rust-openssl/blob/3b064fdb022912bbb98f5b8d9d111aeb6fec8f79/openssl/src/symm.rs#L537

...which is invoked from here
https://github.com/Michael-F-Bryan/libsignal-protocol-rs/blob/bc0d07bec5fd8e853973cdd187ca0d3d7c2618c4/libsignal-protocol/src/crypto/openssl.rs#L70

I have not been able to find out what causes this error. I will try to look more into this but if you have time it would be great if you could have a look as well @Michael-F-Bryan. I think your FFI ninja skillz may be useful here...

To reproduce `cargo test --all --verbose --no-default-features --features="crypto-openssl"`

**Comment by furuholm** [comment]
Here is the error `error:06065064:digital envelope routines:EVP_DecryptFinal_ex:bad decrypt:crypto/evp/evp_enc.c:570`

**Comment by Michael-F-Bryan** [comment]
I'm not able to determine the error just from looking through Travis logs, but that's probably because I'm looking on a phone screen.

When I finish this deployment (Wednesday) to the West Australian fires I'll have my computer, so that should make troubleshooting easier.

Dr what you've seen do far, what do you think the problem might be?

**Comment by furuholm** [comment]
Great to hear that you are helping out with the fires! Just make sure to stay safe...

Found the issue and pushed a fix (4c8e4a0). The same output `Vec` was provided to `Crypter::update`and `Crypter::finalize` which resulted in the data being overwritten.

BTW: Is there a reason to not use the encrypt/decrypt method in https://crates.io/crates/openssl?

**Comment by Michael-F-Bryan** [comment]
> BTW: Is there a reason to not use the encrypt/decrypt functions in https://crates.io/crates/openssl?

I'm not sure. You'd need to ask @shekohex seeing as he did a lot of the OpenSSL work. I'm guessing we needed to do extra customisation that the basic `encrypt()` functions don't support.

What are your thoughts on the whole `serialize()` and `deserialize()` problem? Should we leave some implementations to panic unconditionally, or what about returning a `Result` so we can tell the user an operation isn't supported?

**Comment by furuholm** [comment]
I prefer `Result`. Will change.

**Comment by furuholm** [comment]
I started on this, but I don't like resolving "not implemented" at runtime. How about splitting the `Serializeble` trait into `Serializeble` and `Deserializeble` instead?  (see 8fca321)

**Comment by furuholm** [comment]
@Michael-F-Bryan: have you had a chance to take a look at the change I made in 8fca321?

File: libsignal-protocol/src/lib.rs [hunk 10]
```diff
@@ -133,3 +128,11 @@ pub trait Serializable {
         Ok(())
     }
 }
+
+/// A helper trait for something which can be deserialized from protobufs.
+pub trait Deserializable {
```
Comment (Michael-F-Bryan) [inline libsignal-protocol/src/lib.rs]: Instead of having a `where Self: Sized` bound on `deserialize()` you should declare `Deserializable` as requiring `Sized` (i.e. `trait Deserializable: Sized { ... }`). 

That would line up with [`serde::Deserialize`](https://docs.rs/serde/1.0.104/serde/trait.Deserialize.html). Plus where clauses on trait methods are usually used to let the method opt-out of being object-safe, which doesn't really make sense considering it's got one function which will return an instance of `Self` on the stack.

File: libsignal-protocol/src/macros.rs [hunk 11]
```diff
@@ -1,13 +1,6 @@
 macro_rules! impl_serializable {
     ($name:ty, $serialize:ident, $deserialize:ident) => {
         impl $crate::Serializable for $name {
-            fn deserialize(_data: &[u8]) -> Result<Self, failure::Error>
```
Comment (Michael-F-Bryan) [inline libsignal-protocol/src/macros.rs]: If the `Serializable` trait no longer takes care of `deserialize()`, should we add an `impl_deserializable!()` macro? Otherwise there won't be any way to deserialize anything from this crate that doesn't have a manual `Deserializable` impl.

**Comment by furuholm** [comment]
Hej @Michael-F-Bryan: Sorry for kind of disappearing. I did switch to another solution for the problem I had, but it would be nice to be able to wrap this up. I think you know quite well what you would like to have done. If you want to you can go ahead and fix this. I could try to take a look at it as well but I'm not sure when I will be able to do it...

**Comment by Michael-F-Bryan** [comment]
@furuholm, I don't think I have access to update your PR directly so I'll create a new PR based on top of this one and merge that.

Closed in favour of #57.
